hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java

```diff
@@ -0,0 +1,363 @@
+                                                            DigestType.SHA_384,
+                                                            (short) 1,
+                                                            DataFileCommon.VARIABLE_DATA_SIZE,
+                                                            ks,
+                                                            (short) 1,
+                                                            DataFileCommon.VARIABLE_DATA_SIZE,
+                                                            new OnDiskValueSerializer(md),
+                                                            true));
+
+                            final var label = StateUtils.computeLabel(serviceName, def.stateKey());
+                            hederaState.putServiceStateIfAbsent(md, new VirtualMap<>(label, ds));
+                        }
+                    });
+
+            // Create the writable states. We won't commit anything from these states
+            // until we have completed migration.
+            final var newStates = hederaState.createWritableStates(serviceName);
+
+            // Now we can migrate the schema and then commit all the changes
+            schema.migrate(previousStates, newStates);
+            newStates
+                    .stateKeys()
+                    .forEach(
+                            stateKey ->
+                                    // Reviewers: Should we promote "commit" to WritableKVState?
+                                    ((WritableKVStateBase<?, ?>) newStates.get(stateKey)).commit());
+
+            // And finally we can remove any states we need to remove
+            final var statesToRemove = schema.statesToRemove();
+            statesToRemove.forEach(
+                    stateKey -> hederaState.removeServiceState(serviceName, stateKey));
+        }
+    }
+
+    /**
+     * Given two versions, gets the ordered list of all schemas that must be applied to transition
+     * the merkle tree from some previousVersion to the currentVersion. If {@code previousVersion}
+     * and {@code currentVersion} are the same, then an empty set is returned. In all other cases,
+     * every registered {@link Schema} newer than {@code previousVersion} and less than or equal to
+     * {@code currentVersion} will be returned.
+     *
+     * @param previousVersion The previous version of the merkle tree. May be null for genesis. Must
+     *     be less than or equal to {@code currentVersion}.
+     * @param currentVersion The current version of the application. May NOT be null under any
+     *     condition. Must be greater than or equal to the {@code previousVersion}.
+     * @return An ordered list of {@link Schema}s which, when applied in order, will transition the
+     *     merkle tree from {@code previousVersion} to {@code currentVersion}.
+     */
+    @NonNull
+    private List<Schema> computeApplicableSchemas(
```

Comment:
A bunch of methods here look like they could be static methods. Since this class is really important, consider moving these methods to a static utility class so that this file can remain small.

Reply:
I tend to like keeping them here because I can keep the methods private. If another use case comes up where they need to be reused, we can move them to a utility class (or maybe even make them API on Schema itself!)